src/node_errors.cc

```diff
@@ -843,59 +861,81 @@ void FatalException(Isolate* isolate,
     Abort();
   }
 
+  // Invoke process._fatalException() to give user a chance to handle it.
+  // We have to grab it from the process object since this has been
+  // monkey-patchable.
   Local<Object> process_object = env->process_object();
   Local<String> fatal_exception_string = env->fatal_exception_string();
   Local<Value> fatal_exception_function =
       process_object->Get(env->context(),
                           fatal_exception_string).ToLocalChecked();
-
+  // If the exception happens before process._fatalException is attached
+  // during bootstrap, or if the user has patched it incorrectly, just crash.
   if (!fatal_exception_function->IsFunction()) {
-    // Failed before the process._fatalException function was added!
-    // this is probably pretty bad.  Nothing to do but report and exit.
     ReportException(env, error, message);
     env->Exit(6);
-  } else {
-    errors::TryCatchScope fatal_try_catch(env);
-
-    // Do not call FatalException when _fatalException handler throws
-    fatal_try_catch.SetVerbose(false);
+    return;
+  }
 
+  MaybeLocal<Value> handled;
+  {
+    // We do not expect the global uncaught exception itself to throw any more
+    // exceptions. If it does, crash.
+    errors::TryCatchScope fatal_try_catch(
+        env, errors::TryCatchScope::CatchMode::kFatal);
     Local<Value> argv[2] = { error,
                              Boolean::New(env->isolate(), from_promise) };
 
-    // This will return true if the JS layer handled it, false otherwise
-    MaybeLocal<Value> caught = fatal_exception_function.As<Function>()->Call(
+    handled = fatal_exception_function.As<Function>()->Call(
         env->context(), process_object, arraysize(argv), argv);
+  }
 
-    if (fatal_try_catch.HasTerminated()) return;
-
-    if (fatal_try_catch.HasCaught()) {
-      // The fatal exception function threw, so we must exit
-      ReportException(env, fatal_try_catch);
-      env->Exit(7);
-
-    } else if (caught.ToLocalChecked()->IsFalse()) {
-      ReportException(env, error, message);
+  // The global uncaught exception handler returns true if the user handles it
+  // by e.g. listening to `uncaughtException`. In that case, continue program
+  // execution.
+  // TODO(joyeecheung): This has been only checking that the return value is
+  // exactly false. Investigate whether this can be turned to an "if true"
+  // similar to how the worker global uncaught exception handler handles it.
+  if (!handled.ToLocalChecked()->IsFalse()) {
```

Comment:
I don’t think you can assume `.ToLocalChecked()` here; we also need to return if `handled` is empty.

(As a concrete scenario: I think calling `process.exit()` during a `process.on('uncaughtException')` handler inside a Worker would trigger that code path.)

Reply:
PR with a test that crashes after this change here: #28259

Reply:
Thanks for adding the test!

I added more comments about this. Also left a TODO to return a Maybe here instead because I think at least the `Environment::RunAndClearNativeImmediates` needs to take care of this.